src/codegen/mod.rs

```diff
@@ -1160,35 +1177,57 @@ impl Module {
             .map(Clone::clone)
             .collect();
 
+        tests.into_iter().map(|name| TestCase {
+            name: name.replace("test#", ""),
+            func: self
+                .get_function::<Ctx, fn() -> Verdict<(), ()>>(
+                    name.strip_prefix("pkg.").unwrap(),
+                )
+                .unwrap(),
+        })
+    }
+
+    pub fn run_tests<Ctx: 'static>(&mut self, ctx: Ctx) -> Result<(), ()> {
+        self.run_tests_with_writer(ctx, &mut std::io::stdout())
+    }
+
+    pub fn run_tests_with_writer<Ctx: 'static>(
+        &mut self,
+        mut ctx: Ctx,
+        mut w: &mut impl std::io::Write,
```

Comment:
I think taking `mut w: impl std::io::Write` should be enough. You're taking references elsewhere.